### PR TITLE
Add functionality for grouping and stacking spectrum observations

### PR DESCRIPTION
**Pull request by JouvinLea**
New PR for the grouping after your rebasing.
My test_spectralgrouping doesn't work anymore due to the new read_ogip in spectrum observation but I will try to fix that this morning

This replaces
#464
#429
#381

### CODE REVIEW

**Comment by joleroi** [comment]
Don't worry about the test failures on travis for now. @cdeil is on it (right? :wink: )

File: gammapy/spectrum/spectral_grouping.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import absolute_import, division, print_function, unicode_literals
+import logging
+from astropy.coordinates import Angle
+from astropy.units import Quantity
+import numpy as np
+from ..spectrum.spectrum_extraction import SpectrumObservationList, SpectrumObservation
+from ..data import ObservationGroupAxis, ObservationGroups
+
+__all__ = [
```
Comment (JouvinLea) [inline gammapy/spectrum/spectral_grouping.py]: @joleroi 
Don't know how to add this class to the docstring

File: gammapy/spectrum/spectral_grouping.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            List of SpectrumObservation for each band
+
+        """
+        # Define a new observation table with the number of each band for each observation
+        observation_table = self.spectrum_observation_list.to_observation_table(True)
+        obs_table_grouped = obs_groups.group_observation_table(observation_table)
+
+        nband = obs_groups.n_groups
+        list_band = []
+        for nb in range(nband):
```
Comment (JouvinLea) [inline gammapy/spectrum/spectral_grouping.py]: This loop is taking too much time since there are a lot of empty bands...

**Comment by JouvinLea** [comment]
@joleroi 
Normally the tests are ok now!!
Tell me what you think about what I coded for the grouping!

File: gammapy/spectrum/tests/test_spectralgrouping.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    obs = [23523, 23559, 23526, 23592]
+    store = gammapy_extra.filename("datasets/hess-crab4-hd-hap-prod2")
+    ds = DataStore.from_dir(store)
+
+    ana = SpectrumExtraction(datastore=ds, obs_ids=obs, on_region=on_region,
+                             bkg_method=bkg_method, exclusion=excl,
+                             ebounds=bounds)
+    return ana
+
+def test_define_spectral_groups():
```
Comment (joleroi) [inline gammapy/spectrum/tests/test_spectralgrouping.py]: When running this locally without gammapy-extra I get this error

```
GammapyExtraNotFoundError: The gammapy-extra repo is not available. You have to set the GAMMAPY_EXTRA environment variable to point to the location for it to be found.
```

You have to mark all test that require gammapy-extra with
``
@requires_data('gammapy-extra')
``
see e.g.
https://github.com/gammapy/gammapy/blob/master/gammapy/spectrum/tests/test_spectrum_extraction.py#L17

The same holds for optional dependencies (such as sherpa, and YAML)

File: gammapy/spectrum/tests/test_spectrum_extraction.py [hunk 4]
```diff
@@ -29,17 +57,58 @@ def test_spectrum_extraction(tmpdir):
...
+    assert new_list[1].obs_id == 23592
+
+@requires_data('gammapy-extra')
+def test_spectrum_extraction_grouping_from_an_observation_list():
+    """
+    Test the stacking for two observations
+    """
+
+    ana = make_spectrum_extraction()
+    ana.extract_spectrum()
```
Comment (joleroi) [inline gammapy/spectrum/tests/test_spectrum_extraction.py]: Here I get
```
E   KeyError: u'MUONEFF'
```
I suppose you have not commited your changed eventlist to ``gammapy-extra``

File: gammapy/spectrum/spectral_grouping.py (hunk 1 above)
Comment (joleroi) [inline gammapy/spectrum/spectral_grouping.py]: You have to modify the file ``gammapy/spectrum/_init__py``. Put ``from_spectral_grouping import *``

File: gammapy/spectrum/spectral_grouping.py (hunk 2 above)
Comment (joleroi) [inline gammapy/spectrum/spectral_grouping.py]: I also don't know. Let's worry about this later, ok?

File: gammapy/spectrum/spectrum_extraction.py [hunk 5]
```diff
@@ -669,13 +708,19 @@ def read_ogip(cls, dir='ogip_data'):
...
-        """Create `~gammapy.data.ObservationTable`"""
-        names = ['OBS_ID', 'PHAFILE', 'OFFSET']
+    def to_observation_table(self, moreparameters=False):
+        """Create `~gammapy.data.ObservationTable"""
+        names = ['OBS_ID', 'PHAFILE']
         col1 = [o.obs_id for o in self]
         col2 = [o.meta.phafile for o in self]
-        col3 = [o.meta.offset.value for o in self]
-        return ObservationTable(data=[col1, col2, col3], names=names)
+        if moreparameters:
```
Comment (joleroi) [inline gammapy/spectrum/spectrum_extraction.py]: If you want to you can modify this to write all info that exist in meta. something like

```
keys = ['phafile', 'muon_eff', ....]
for key in keys:
   if key in meta:
       write
```

File: gammapy/spectrum/spectrum_fit.py [hunk 6]
```diff
@@ -293,12 +290,13 @@ def _run_sherpa_fit(self):
...
 
+        namedataset = []
         for i in range(len(ds.datasets)):
             datastack.notice_id(i + 1, thres_lo[i], thres_hi[i])
-
+            namedataset.append(i + 1)
         datastack.set_stat(self.statistic)
-        ds.fit()
-        datastack.covar()
+        ds.fit(*namedataset)
```
Comment (joleroi) [inline gammapy/spectrum/spectrum_fit.py]: Just for curiosity, since I don't really know Sherpa. What does this do?

**Comment by joleroi** [comment]
Hi,

nice work! Let's get it into master soon. I have left some minor comments.
Once thing you could to is remove the file
``gammapy/spectrum/spectrum_grouping.py``
it is empty
``git rm gammapy/spectrum/spectrum_grouping.py``

I will start using the grouping code and see what I can add (the energy dependent livetime e.g.)

**Comment by joleroi** [comment]
FYI: I will close

https://github.com/gammapy/gammapy/pull/464
https://github.com/gammapy/gammapy/pull/429
https://github.com/gammapy/gammapy/pull/381

File: gammapy/spectrum/spectrum_fit.py (hunk 6 above)
Comment (JouvinLea) [inline gammapy/spectrum/spectrum_fit.py]: I encountered some problems with sherpa when in the same script you fit multiples times. For example, when I tested the grouping I fit multiple times for different groups and sherpa if you don't add this option keep in memory the others dataset so it gives you each time the result for the different datasets that you used in your script. This line allows to force sherpa during the fit to take only into account the dataset that you gives with the ogip files. I'm not sure if this is clear...

Comment (joleroi) [inline gammapy/spectrum/spectrum_fit.py]: I get the idea. At some point we should move to the real sherpa class interface anyways, so problems like this should go away ..

File: gammapy/spectrum/spectrum_extraction.py [hunk 7]
```diff
@@ -669,15 +710,20 @@ def read_ogip(cls, dir='ogip_data'):
...
-        """Create `~gammapy.data.ObservationTable`"""
-        names = ['OBS_ID', 'PHAFILE', 'OFFSET']
-        col1 = [o.obs_id for o in self]
-        col2 = [o.meta.phafile for o in self]
-        col3 = [o.meta.offset.value for o in self]
-        return ObservationTable(data=[col1, col2, col3], names=names)
-
-
+    def to_observation_table(self, moreparameters=False):
+        """Create `~gammapy.data.ObservationTable"""
```
Comment (JouvinLea) [inline gammapy/spectrum/spectrum_extraction.py]: seems better now?

File: gammapy/spectrum/spectral_grouping.py (hunk 1 above)
Comment (JouvinLea) [inline gammapy/spectrum/spectral_grouping.py]: @joleroi 
I added this line in the init:
from .spectral_grouping import *

But when I run the python setup.py build_sphinx, I got that error:
```
  File "/Users/jouvin/gammapy/build/lib.macosx-10.9-x86_64-2.7/gammapy/spectrum/SpectralGrouping.py", line 25
    +                               ZenRange=[0., 70.], NZenbin=30):
    ^
SyntaxError: invalid syntax
Sphinx Documentation subprocess failed with return code 1

```

Comment (cdeil) [inline gammapy/spectrum/spectral_grouping.py]: Try: ``make clean``
or ``python -c 'import gammapy.spectrum'``
to see if there really is a SyntaxError in that file.

Comment (JouvinLea) [inline gammapy/spectrum/spectral_grouping.py]: Thanks

File: gammapy/spectrum/spectrum_extraction.py (hunk 7 above)
Comment (joleroi) [inline gammapy/spectrum/spectrum_extraction.py]: :+1:

**Comment by joleroi** [comment]
You have updated ``gammapy-extra``, right? 
I still get some error when I run the test, but I guess some of them are due to errors I did when doing the rebase. If you think we're ready, I will merge and then fix the remaining tests myself.

**Comment by JouvinLea** [comment]
Yeah normally Christoph merged the gammapy-extra with the keyword MUONEFF, do you still get an error on that?

**Comment by joleroi** [comment]
I get ``E           KeyError: u'muoneff_MIN'``

**Comment by cdeil** [comment]
@JouvinLea - Can you reproduce the KeyError locally?
If not, maybe you have modified files in gammapy-extra?

I've restarted the CI builds just now.

This should be merged today so that further development isn't blocked for next week.
f test errors remain, please put `@pytest.mark.xfail` for now on the test, but still merge.

**Comment by cdeil** [comment]
@joleroi - Can you please integrate this PR first?

The possible workflows are:
1. Merge into master and then continue there or in another pull request.
2. Continue with this branch and make your own pull request containing the commits here.

If there's some stuff to do, 2. might be better, because then you get travis-ci testing and can review the diff before merging into master.

**Comment by cdeil** [comment]
@joleroi - This PR can be closed, right?

**Pull request closed without merging**